sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -648,34 +612,44 @@ message Environment {
   string url = 1;
 }
 
-// Description of a function in a Beam pipeline.
+// A specification of a user defined function.
 //
-// Contains one of _or both of_ a UrnWithParameter specifying the function
-// and the specification for how to execute it against a particular
-// SDK's harness.
 message FunctionSpec {
 
-  // (Optional) An SDK-independent specification of this function.
-  // If present, this must _fully_ specify the function.
+  // (Required) A full specification of this function.
+  //
+  // There are two main styles, which need not be fully distinct
   //
-  // For example the distinguished urn "urn:beam:windowfn:FixedWindows" with
-  // payload `{ duration: n }` fully specifies a windowing function which can
-  // be implemented by the SDK constructing the pipeline, by another SDK (for
-  // language-to-language fusion compatibility) or by the runner directly.
+  // 1. A well-known URN like "urn:beam:windowfn:FixedWindows"
+  //    with an agreed-upon payload (e.g. a duration) that the
+  //    SDK generating this pipeline understands, but is also
+  //    possibly understood by a runnner and/or other SDKs.
+  // 2. A URN that the runner does not understand, such as
+  //    "urn:beam:dofn:javasdk:1.0" that might indicate that
+  //    it is just a serialized Java DoFn from a particular
+  //    version of the Beam Java SDK. The payload will often
+  //    then be an opaque message such as bytes in a
+  //    language-specific serialization format.
+  //
+  // In both cases, the SDK generating the pipeline understands
+  // the data attached to the URN.
   UrnWithParameter spec = 1;
 
-  // (Optional) An SDK-specific specification for how to execute this function,
-  // including a specification of the environment in which the function
-  // can be interpreted and executed.
-  SdkFunctionSpec sdk_fn_spec = 2;
+  // (Required) Reference to the specification of the execution environment
+  // required to invoke this function.
```

Comment:
"required to invoke this function" seems to suggest the wrong thing (it is the only environment capable of invoking the function). "an execution environment capable of invoking this function" seems more like what this actually is, unless I've misunderstood something.